#61 adds api keys as authentication method

LeoTuet opened the pull request.

LeoTuet committed:
#61 adds api keys as authentication method

LeoTuet committed:
#61 corrects api key controller import

94726 commented on src/repositories/schemas.ts:
```diff
@@ -34,3 +34,9 @@ export interface Admin {
+
+export interface ApiKey {
+  _id: string;
+  name: string;
```
I guess the name is supposed to be unique. Please add it as an index field in mongoDB.
Not sure whether the key should also be unique and part of the index.

see
https://github.com/LeoTuet/backend-we-want-to-sleep/blob/519800d247c57c571baca26ce0b3a431dcb8a490/src/repositories/connectToDB.ts#L23-L30

marcschwaiger replied:
I think saving the name of the admin that created the key would make sense here. 
See createdBy in Ballot or Token.

94726 commented on src/handler/ApiKeyHandler.ts:
```diff
@@ -0,0 +1,22 @@
+const apiKeyService = new ApiKeyService();
+
+export class ApiKeyHandler {
+  public async addApiKey(name: string, key: string) {
+    if (await apiKeyService.checkIfApiKeyNameExists(name))
+      throw new UnprocessableEntity(
+        "A api key with the given name already exists"
+      );
+
+    await apiKeyService.addApiKey(name, key);
```
Mind that the key itself should probably also be unique.

94726 reviewed (changes requested):
I think that an api key should not be human-made, but rather be generated by the backend.
Automatically generated keys could improve the security.

94726 commented:
Just noticed, could you please hash the api key before dumping it into the database.

marcschwaiger commented on src/routes/index.ts:
```diff
@@ -47,5 +48,10 @@ const authRouter = Router();
 authRouter.post("/login", adminLoginLimiter, AuthController.login);
 
+const apiKeyRouter = Router();
+baseRouter.use("/apiKey", apiKeyRouter);
```
```suggestion
baseRouter.use("/api-key", apiKeyRouter);
```
Please don't use uppercase letters in url paths

LeoTuet replied:
I used uppercase because of [ballotID](https://github.com/LeoTuet/backend-we-want-to-sleep/pull/62/files/7bc5f5a811b787a1a3115210e016d706dc4f23aa#diff-4c1a276a826a5147689418cdbf9bde24c8326cedb30a61144fe7867bd99113efR31)

LeoTuet replied:
I used uppercase because [ballotID](https://github.com/LeoTuet/backend-we-want-to-sleep/blob/main/src/routes/index.ts#L20) is also uppercase.

marcschwaiger replied:
[ballotID](https://github.com/LeoTuet/backend-we-want-to-sleep/blob/main/src/routes/index.ts#L20) is a variable and not in the path

LeoTuet replied:
oh f... overlooked that

LeoTuet committed:
#61 changes to generated api keys and hashing

LeoTuet committed:
#61 adds crypto-js as new dependency

LeoTuet committed:
#61 adds types for crypto-js

marcschwaiger commented on src/services/ApiKeyService.ts:
```diff
@@ -0,0 +1,34 @@
+  public async addApiKey(name: string, createdBy: string) {
+    const key = this.generateApiKey();
+
+    if (this.checkIfApiKeyExists(key)) this.addApiKey(name, createdBy);
```
You forgot two awaits and one return here

marcschwaiger commented on src/services/ApiKeyService.ts:
```diff
@@ -0,0 +1,34 @@
+
+    if (this.checkIfApiKeyExists(key)) this.addApiKey(name, createdBy);
+
+    const keyHash = sha256(key).toString();
```
we talked about using bcrypt and prepending the name in front of the key

94726 replied:
But that is just an uncommen implementation feeling a bit to amogus

LeoTuet replied:
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣠⣴⣶⣿⣿⣷⣶⣄⣀⣀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⣰⣾⣿⣿⡿⢿⣿⣿⣿⣿⣿⣿⣿⣷⣦⡀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⢀⣾⣿⣿⡟⠁⣰⣿⣿⣿⡿⠿⠻⠿⣿⣿⣿⣿⣧⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⣾⣿⣿⠏⠀⣴⣿⣿⣿⠉⠀⠀⠀⠀⠀⠈⢻⣿⣿⣇⠀⠀⠀
⠀⠀⠀⠀⢀⣠⣼⣿⣿⡏⠀⢠⣿⣿⣿⠇⠀⠀⠀⠀⠀⠀⠀⠈⣿⣿⣿⡀⠀⠀
⠀⠀⠀⣰⣿⣿⣿⣿⣿⡇⠀⢸⣿⣿⣿⡀⠀⠀⠀⠀⠀⠀⠀⠀⣿⣿⣿⡇⠀⠀
⠀⠀⢰⣿⣿⡿⣿⣿⣿⡇⠀⠘⣿⣿⣿⣧⠀⠀⠀⠀⠀⠀⢀⣸⣿⣿⣿⠁⠀⠀
⠀⠀⣿⣿⣿⠁⣿⣿⣿⡇⠀⠀⠻⣿⣿⣿⣷⣶⣶⣶⣶⣶⣿⣿⣿⣿⠃⠀⠀⠀
⠀⢰⣿⣿⡇⠀⣿⣿⣿⠀⠀⠀⠀⠈⠻⣿⣿⣿⣿⣿⣿⣿⣿⣿⠟⠁⠀⠀⠀⠀
⠀⢸⣿⣿⡇⠀⣿⣿⣿⠀⠀⠀⠀⠀⠀⠀⠉⠛⠛⠛⠉⢉⣿⣿⠀⠀⠀⠀⠀⠀
⠀⢸⣿⣿⣇⠀⣿⣿⣿⠀⠀⠀⠀⠀⢀⣤⣤⣤⡀⠀⠀⢸⣿⣿⣿⣷⣦⠀⠀⠀
⠀⠀⢻⣿⣿⣶⣿⣿⣿⠀⠀⠀⠀⠀⠈⠻⣿⣿⣿⣦⡀⠀⠉⠉⠻⣿⣿⡇⠀⠀
⠀⠀⠀⠛⠿⣿⣿⣿⣿⣷⣤⡀⠀⠀⠀⠀⠈⠹⣿⣿⣇⣀⠀⣠⣾⣿⣿⡇⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠹⣿⣿⣿⣿⣦⣤⣤⣤⣤⣾⣿⣿⣿⣿⣿⣿⣿⣿⡟⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠉⠻⢿⣿⣿⣿⣿⣿⣿⠿⠋⠉⠛⠋⠉⠉⠁⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠉⠉⠉⠁

marcschwaiger commented on src/services/ApiKeyService.ts:
```diff
@@ -0,0 +1,34 @@
+
+  public deleteApiKey = ApiKeyRepository.deleteApiKey;
+
+  public generateApiKey() {
```
I think exporting this function could lead to confusion. See [TokenService#generateUnformattedToken](https://github.com/LeoTuet/backend-we-want-to-sleep/blob/main/src/services/TokenService.ts#L4) and call it generateUnformattedApiKey?

marcschwaiger commented on src/handler/AuthHandler.ts:
```diff
@@ -1,9 +1,11 @@
 import { AuthService } from "../services/AuthService";
 
 const authService = new AuthService();
+const apiKeyService = new ApiKeyService()
```
Please format your code using prettier 
(You forgot the ; )

marcschwaiger commented on src/handler/AuthHandler.ts:
```diff
@@ -25,6 +27,12 @@ export class AuthHandler {
+  async authenticateWithApiKey(key: string): Promise<boolean> {
+    if (!(await apiKeyService.checkIfApiKeyExists(key)))
+      throw new Unauthorized("API key is not valid");
+    return true;
```
You can drop the return value here as you dont even need it in [AuthMiddleware](https://github.com/LeoTuet/backend-we-want-to-sleep/blob/61-authentication-for-metrics/src/middleware/AuthMiddleware.ts#L19)

LeoTuet committed:
#62 implements comments from pr

marcschwaiger committed:
#62 Minor improvements to pr

- Fixed collision handling logic in api key service
- Changed api key length to 32
- Refactored

Signed-off-by: Schwaiger Marc <marcschwaiger@outlook.de>

94726 commented on src/controller/ApiKeyController.ts:
```diff
@@ -0,0 +1,31 @@
+    );
+    res.json({
+      data: {
+        added: {
+          name: req.body.name,
+          key: apiKey,
+        },
```
Isn't that `added` object kinda unnecessary?

94726 commented on src/middleware/AuthMiddleware.ts:
```diff
@@ -6,10 +6,21 @@ const authHandler = new AuthHandler();
+export const hasValidApiKey = asyncHandler(
+  async (req: Request<{}, {}, {}>, res, next: NextFunction) => {
+    await authHandler.authenticateWithApiKey(
+      req.headers["x-api-key"] as string
```
Type assertions should be avoided at all cost.
```suggestion
      req.headers["x-api-key"].toString()
```

94726 commented on src/repositories/connectToDB.ts:
```diff
@@ -25,6 +25,7 @@ async function prepareDatabase() {
     { token: 1, ballot: 1 },
     { unique: true }
   );
+  await getCollection("api-key").createIndex({ name: 1 }, { unique: true });
```
Please also add the `key` field as part of the collection-index as it's being used to check the existence of incoming requests with api-keys:
https://github.com/LeoTuet/backend-we-want-to-sleep/blob/15d45eb20491be6fc2ddf5ab5c04687d4644f497/src/handler/AuthHandler.ts#L30-L33

94726 commented on src/routes/index.ts:
```diff
@@ -47,5 +48,10 @@ const authRouter = Router();
 
+const apiKeyRouter = Router();
+baseRouter.use("/api-key", apiKeyRouter);
+apiKeyRouter.post("/", isAdmin, APIKeyController.add);
+apiKeyRouter.delete("/", isAdmin, APIKeyController.delete);
```
A `get` endpoint might be handy.

94726 reviewed (changes requested):
Nice do while loop btw.
https://github.com/LeoTuet/backend-we-want-to-sleep/blob/15d45eb20491be6fc2ddf5ab5c04687d4644f497/src/services/ApiKeyService.ts#L15-L16

marcschwaiger committed:
#62 Implemented changes from pr comments

- Added an endpoint to get all api keys
- Renamed routes\index.ts to router.ts

Signed-off-by: Schwaiger Marc <marcschwaiger@outlook.de>

marcschwaiger committed:
#62 Fix for commit 9674fd7

Signed-off-by: Schwaiger Marc <marcschwaiger@outlook.de>

94726 commented on src/handler/ApiKeyHandler.ts:
```diff
@@ -1,10 +1,22 @@
+  delete key.keyHash;
+  delete key._id;
+  return key;
+};
+const convertApiKeys = (keys: ApiKey[]): ApiKeyInfo[] => {
```
```suggestion
};

const convertApiKeys = (keys: ApiKey[]): ApiKeyInfo[] => {
```

marcschwaiger committed:
#62 Added spaces between converter methods

Signed-off-by: Schwaiger Marc <marcschwaiger@outlook.de>
